### ISSUE TITLE
_beam_search_ops.so and PyFunc don't work with TF Serving (Was: In-Graph GatherTree)

### ISSUE THREAD

**Issue by baobao7766** [tensorflow/tensorflow#22133]
Hi, i want to implement `tensorflow/contrib/seq2seq/python/ops/beam_search_ops.py >> gather_tree()`, because of `_beam_search_ops.so` and `PyFunc` are not worked in tf-serving, so is there any implementation of GatherTree by while_loop?

**Comment by baobao7766** [comment]
I am using `python2.7 + tensorflow-1.2.1 + tf-serving-1.10`. Develop with `macOS + tensorflow-cpu-1.2.1`.
And the project is a seq2seq use beamSearch decoder.

**Comment by ebrevdo** [comment]
This is gonna be super slow.  I'm going to rename this bug to "_beam_search_ops.so and PyFunc don't work with TF Serving".

**Comment by ebrevdo** [comment]
Can you provide a simple repro for trying to load this in TF Serving so we can check the failure?

**Comment by ebrevdo** [comment]
@martinwicke @qlzh727 we may be able to resolve this by moving the c++ kernels in tf.contrib.seq2seq to core.

**Comment by guillaumekln** [comment]
TensorFlow Serving already links against `tensorflow/contrib:contrib_kernels`:

https://github.com/tensorflow/serving/blob/r1.10/tensorflow_serving/model_servers/BUILD#L260L263

However, the `beam_search_ops` kernels were added to this rule in TensorFlow 1.3.

**Comment by baobao7766** [comment]
Thank you for your reply, so if i re-train the model with higher version, this will be solved ? @guillaumekln 
The code is what i do and my solution in project. @ebrevdo (I'm sorry for my terrible code ^_^)
`
beam_decoder = BeamSearchDecoder(
    cell=dec_cell,
    embedding=emb_dec,
    start_tokens=tf.ones_like(qes_seq_len) * go_id, 
    end_token=eos_id,
    initial_state=init_state,
    beam_width=beam_width,
    output_layer=output_layer,
    diversity_promoting=diversity_promoting,
    length_penalty_weight=0.6)
beam_output_idx, beam_output_state, beam_lengths = tf.contrib.seq2seq.dynamic_decode(
    decoder=beam_decoder,
    maximum_iterations=max_decode_len)

with tf.variable_scope("beamSearch"):
    beam_out = index2symbol.lookup(
        tf.cast(beam_output_idx.predicted_ids, tf.int64),
        name='beam_out')
    beam_scores = beam_output_idx.beam_search_decoder_output[0]
    beam_log_probs = beam_output_state[1]

    beam_out_trans = tf.transpose(beam_out, perm=[0, 2, 1])
    beam_scores_trans = tf.transpose(beam_scores, perm=[0, 2, 1])
......
out_infers  = tf.saved_model.utils.build_tensor_info(
                tensor=graph.get_tensor_by_name("beamCandidate/chosenAnswers/Merge:0"))
out_score = tf.saved_model.utils.build_tensor_info(
                tensor=graph.get_tensor_by_name("beamCandidate/chosenAnswers/Merge_1:0"))
prediction_signature = tf.saved_model.signature_def_utils.build_signature_def(
    inputs={
        'question': question_placeholder,
        'answer': answer_placeholder,
        'question_len': qes_len_placeholder,
        'answer_len': ans_len_placeholder,
        'answer_threshold': answer_threshold,
    },
    outputs={
        'output': out_infers,
        'score': out_score,
    },
    method_name=tf.saved_model.signature_constants.PREDICT_METHOD_NAME)
`
This is my solution, and it's not super slow at all, but it will be nice if `_beam_search_ops` worked in higher version, i will take this step later.
`
def gather_tree_tf(step_ids, parent_ids, sequence_lengths):
    with tf.variable_scope(name_or_scope="GatherTree", reuse=None):
        idx_shape = tf.shape(step_ids)      # 4, 2, 3
        idx_dtype = step_ids.dtype

        max_time    = idx_shape[0]  # 4
        batch_size  = idx_shape[1]  # 2
        beam_width  = idx_shape[2]  # 3

        init_res = tf.zeros(idx_shape, dtype=idx_dtype)
        init_cnt = tf.constant(0)

        a = tf.reshape(tf.tile(tf.expand_dims(tf.range(beam_width), axis=1), [1, batch_size]), [-1, ])
        b = tf.tile(tf.range(batch_size), [beam_width])
        idx_arr = tf.concat(
            [
                tf.expand_dims(a, axis=1),  # beam_width
                tf.expand_dims(b, axis=1),  # batch_size
            ], axis=-1)
        max_cnt = tf.shape(idx_arr)[0]  # 8

        def condition(cnt, res_arr):
            return tf.less(cnt, max_cnt)

        def body(cnt, res_arr):
            beam_id = idx_arr[cnt][0]   # 0
            batch   = idx_arr[cnt][1]   # 0

            max_len = sequence_lengths[batch][beam_id]  # 4
            parent = parent_ids[max_len - 1][batch][beam_id]    # 2
            temp_idx = tf.concat([tf.expand_dims(batch, axis=0), tf.expand_dims(beam_id, axis=0)], axis=-1)
            temp_a = tf.expand_dims(tf.range(start=max_len-1, limit=max_time), axis=1)
            temp_b = tf.tile(tf.expand_dims(temp_idx, axis=0), [max_time-max_len+1, 1])
            mask_idx = tf.concat([temp_a, temp_b], axis=1)
            mask = indice2mask3D(idxs=mask_idx, shape=idx_shape)
            new_res_arr = tf.where(mask, step_ids, res_arr)

            def condition2(cnt2, parent2, res_arr2):
                return tf.greater(cnt2, 0)

            def body2(cnt2, parent2, res_arr2):
                level = cnt2 - 1
                new_res_arr2 = tf.reshape(
                    tf.tile(tf.expand_dims(step_ids[level][batch][parent2], axis=0), [tf.reduce_prod(idx_shape)]),
                    idx_shape)
                temp_idx2 = tf.expand_dims(tf.concat(
                    [
                        tf.expand_dims(level, axis=0),  # time = 2
                        tf.expand_dims(batch, axis=0),  # batch = 0
                        tf.expand_dims(beam_id, axis=0),  # parent_beam = 2, beam_id = 0
                    ], axis=0), axis=0)
                mask2 = indice2mask3D(idxs=temp_idx2, shape=idx_shape)
                new_res_arr2 = tf.where(mask2, new_res_arr2, res_arr2)
                new_parent = parent_ids[level][batch][parent2]
                return tf.subtract(cnt2, 1), new_parent, new_res_arr2

            out_cnt2, out_new_parent2, out_res_arr2 = tf.while_loop(
                cond=condition2,
                body=body2,
                loop_vars=[max_len-1, parent, new_res_arr],
                name="SubLoop2")
            new_cnt = tf.add(cnt, 1, name="newCount")
            return new_cnt, out_res_arr2

        final_cnt, final_res_arr = tf.while_loop(
            cond=condition,
            body=body,
            loop_vars=[init_cnt, init_res],
            name="MainLoop")
        return final_res_arr

def indice2mask3D(idxs, shape):
    with tf.variable_scope("Mask3D"):
        t = tf.range(tf.reduce_prod(shape))
        max_cnt = tf.shape(idxs)[0]
        init_mask = tf.zeros(shape=shape, dtype=tf.bool)
        def cond(cnt, mask):
            return tf.less(cnt, max_cnt)
        def body(cnt, mask):
            idx = idxs[cnt]
            pos = idx[0] * shape[1] * shape[2] + idx[1] * shape[2] + idx[2]
            cur_mask = tf.reshape(tf.equal(t, pos), shape)
            return tf.add(cnt, 1), tf.logical_or(mask, cur_mask)
        final_cnt, final_mask = tf.while_loop(
            cond=cond, body=body, loop_vars=[0, init_mask],
            name="ResultLoop")
        return final_mask
`

**Comment by guillaumekln** [comment]
I had success using `BeamSearchDecoder` in Serving with TensorFlow 1.4 and above.

**Comment by baobao7766** [comment]
Nice, thank you very much! And it's time to update my tensorflow.
